### ISSUE TITLE
Avoiding cast of certain tensors in in the optimizer state dict when load_state_dict() is called

### ISSUE THREAD

**Issue by mikaylagawarecki** [pytorch/pytorch#75224]
### 🐛 Describe the bug

In #71333, #71334 and #71335 certain items in the state dict for various optimizers were changed from floats/ints to singleton Torch.Tensors (most notably `state['step']`). The rationale was to consolidate all updates to the optimizer state to the the functional API provided in [`torch/optim/_functional.py`](https://github.com/pytorch/pytorch/blob/master/torch/optim/_functional.py) which is called by `Optimizer.step()` and hence remove snippets like the following where some updates were done in` Optimizer.step() `and some were done in the corresponding call to the functional API

https://github.com/pytorch/pytorch/blob/f0e972a481b51ddf9d21d3ba8e8ef0ec2e15cd8a/torch/optim/_multi_tensor/nadam.py#L87-L105

#74424 surfaced that this might lead to regressions in optimizer performance where `state['step']` was mistakenly casted to CUDA/XLA as `load_state_dict()` casts all tensors in the state_dict to the device of the param.

https://github.com/pytorch/pytorch/blob/c064ea07f6263090a40b1a6fc5fe5973a76785e5/torch/optim/optimizer.py#L154-L179

#75214 avoids the cast for state['step'] to unblock XLA, follow up work is to determine whether the other items that were moved to tensors caused regressions/whether it would be feasible to avoid the cast for all singleton tensors in the state dict. Creating this issue to track

### Versions

N/A

cc @ezyang @gchanan @zou3519 @vincentqb @jbschlosser @albanD

**Comment by gchanan** [comment]
I'm not sure if I'm reading this correctly -- is `state[step]` floating point?  Or what is `param` here?

**Comment by mikaylagawarecki** [comment]
`param` is the parameters that the optimizer is optimizing (like the weights). If the `param` is floating point all tensors in the optimizer state dict are casted to the device and dtype of the param, so changing `state['step']` to a tensor caused it to be cast to XLA/CUDA

**Comment by jbschlosser** [comment]
Besides `step`, I found the following singleton tensors in the state for the `torch.optim` optimizers:

* `ASGD`: `eta` and `mu`
    * `eta` is [accessed](https://github.com/pytorch/pytorch/blob/3ac6106523707b56ac8124391106638309c397d7/torch/optim/asgd.py#L254) via `_get_value()`, which results in an `item()` call
    * `mu` is [accessed](https://github.com/pytorch/pytorch/blob/3ac6106523707b56ac8124391106638309c397d7/torch/optim/asgd.py#L262) via an `item()` call
* `LBFGS`: `H_diag` and the list `al`
    * `H_diag` is used in a [`torch.mul()` call](https://github.com/pytorch/pytorch/blob/3ac6106523707b56ac8124391106638309c397d7/torch/optim/lbfgs.py#L387) with a tensor on the same device as the parameters, but this is no problem since it's a singleton tensor
    * `al[i]` [computations](https://github.com/pytorch/pytorch/blob/3ac6106523707b56ac8124391106638309c397d7/torch/optim/lbfgs.py#L382) will be on the same device as the parameters, but there's no problem if this isn't the case
* `NAdam`: `mu_product`
    * `mu_product` is [accessed](https://github.com/pytorch/pytorch/blob/3ac6106523707b56ac8124391106638309c397d7/torch/optim/nadam.py#L266) via `_get_value()`, which results in an `item()` call

It's likely the `item()` calls would cause regressions on CUDA. I opened #91454 to keep these tensors on CPU and it passes CI.

@mikaylagawarecki / @albanD One thing I'm not clear on: why is `step` [left out from being casted to the param's dtype](https://github.com/pytorch/pytorch/blob/3ac6106523707b56ac8124391106638309c397d7/torch/optim/optimizer.py#L343-L345)? Does it make sense to do this for all singleton tensors? In general, they are used in math involving the parameters, so an argument could be made their dtype should match that of the parameters. On the other hand, for low-precision parameters, the accuracy of the singleton tensor calculations may suffer.

**Comment by mikaylagawarecki** [comment]
@jbschlosser The context of leaving `step` out from being casted to param dtype/device was to fix this issue https://github.com/pytorch/pytorch/issues/74424 (more so for device rather than dtype). Since `step` should always be a whole number I think casting/not casting it to the param's floating point dtype would not cause accuracy issues, I'm not sure whether this is the case for other singleton tensors
